test/matrix_test.go

```diff
@@ -0,0 +1,497 @@
+	defer tearDown(ctx, t, c, namespace)
+	t.Logf("Creating Task in namespace %s", namespace)
+	task := parse.MustParseV1Task(t, fmt.Sprintf(`
+metadata:
+  name: mytask
+  namespace: %s
+spec:
+  params:
+    - name: exit-code
+  steps:
+    - name: echo
+      image: alpine
+      script: |
+        exit "$(params.exit-code)"
+`, namespace))
+
+	if _, err := c.V1TaskClient.Create(ctx, task, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create Task `%s`: %s", task.Name, err)
+	}
+
+	pipelineRun := parse.MustParseV1PipelineRun(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  serviceAccountName: "default"
+  pipelineSpec:
+    tasks:
+      - name: exit-with-zero-and-non-zero
+        taskRef:
+          name: mytask
+        matrix:
+          params:
+          - name: exit-code
+            value:
+              - "0"
+              - "0"
+              - "1"
+`, helpers.ObjectNameForTest(t), namespace))
+	prName := pipelineRun.Name
+
+	t.Logf("Creating PipelineRun %s", prName)
+	if _, err := c.V1PipelineRunClient.Create(ctx, pipelineRun, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create Pipeline `%s`: %s", prName, err)
+	}
+
+	t.Logf("Waiting for PipelineRun in namespace %s to fail", namespace)
+	if err := WaitForPipelineRunState(ctx, c, prName, timeout, PipelineRunFailed(prName), "PipelineRunFailed", v1Version); err != nil {
+		t.Errorf("Error waiting for PipelineRun to finish: %s", err)
+	}
+
+	pr, err := c.V1PipelineRunClient.Get(ctx, prName, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Couldn't get expected PipelineRun %s: %s", pr.Name, err)
+	}
+
+	if pr.Status.GetCondition(apis.ConditionSucceeded).IsTrue() {
+		t.Errorf("Expected PipelineRun to fail but found condition: %s", pr.Status.GetCondition(apis.ConditionSucceeded))
+	}
+
+	taskRun1 := parse.MustParseV1TaskRun(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  params:
+  - name: exit-code
+    value: 0
+  serviceAccountName: default
+  timeout: 1h0m0s
+  taskRef:
+    name: mytask
+    kind: Task
+`, "taskrun1", namespace))
+
+	taskRun2 := parse.MustParseV1TaskRun(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  params:
+  - name: exit-code
+    value: 0
+  serviceAccountName: default
+  timeout: 1h0m0s
+  taskRef:
+    name: mytask
+    kind: Task
+`, "taskrun2", namespace))
+
+	taskRun3 := parse.MustParseV1TaskRun(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  params:
+  - name: exit-code
+    value: 1
+  serviceAccountName: default
+  timeout: 1h0m0s
+  taskRef:
+    name: mytask
+    kind: Task
+`, "taskrun3", namespace))
+
+	expectedTaskRuns := []*v1.TaskRun{taskRun1, taskRun2, taskRun3}
+
+	actualTaskrunList, err := c.V1TaskRunClient.List(ctx, metav1.ListOptions{})
+	if err != nil {
+		t.Fatalf("Error listing TaskRuns for PipelineRun %s: %s", prName, err)
+	}
+
+	for i := range actualTaskrunList.Items {
```

Comment:
It looks like this test relies on the order taskruns are returned from the fake client. can you update it to sort the slices before comparing? e.g. with https://pkg.go.dev/github.com/google/go-cmp/cmp/cmpopts#SortSlices
Same comment for the previous test class

Reply:
Added sorting

Reply:
I think the idea behind sortSlices here would be that you use it to compare actualTaskRunList and expectedTaskRuns directly. As used here, I think you're sorting the params in the taskrun by name (which is probably still valuable) but still relying on the order of taskruns returned by the fake client.